feat: lost keycard flow - convert keycard account to regular

igor-sirotin opened the pull request:
Fixes https://github.com/status-im/status-desktop/issues/17284

# Description

Nim-side:

1. Implemented `LoginWithLostKeycardSeedphrase`
2. Implemented `KeycardConvertAccountTask`
3. Added handling of keycard account conversion result 
4. Added handling of re-encryption signals
5. Introduce `moment` for post-onboarding tasks
To distinguish, should one be executed after login or when main module loaded.
6. Implemented login using mnemonic
7. Re-implement save biometrics as a post-onboarding task

QML:

1. Implemented `ConvertKeycardAccountPage` (with an according storybook page)
   - This requires conversation with design team.
2. Added re-ecnryption splash screen
   - This requires conversation with design team. I reused the standard `SplashScreen` for this. 
   - I'm not sure if the way I used `StackView` is correct, perhaps there is a better approah.
3. Added `infiniteLoading` option for `SplashScreen`.
   - This is used for the re-encryption screen.

# Screenshot

## Pre-step: create keycard account (with biometrics enabled)

https://github.com/user-attachments/assets/bdf57fae-f6fe-4576-96d8-b00eb713aa14

## Convert keycard account to regular

https://github.com/user-attachments/assets/2d96420b-10bd-4f2d-ae1a-19920c6ab466

caybro commented on ui/main.qml:
```diff
@@ -480,6 +486,12 @@ StatusWindow {
                     return
                 }
+
+                if (flow === Onboarding.OnboardingFlow.LoginWithLostKeycardSeedphrase) {
```
I think the progress page push should not happen here but not sure about the correct place where to put it... :thinking: any ideas @micieslak ?

caybro replied:
I think this should happen earlier, inside OnboardingLayout or Flow; and then we don't emit the finished signal. The result is that we show the progress page and then Quit anyway right?

micieslak replied:
Yeah, I agree that putting into `OnboardingLayout` is good idea. Ideally we should not modify the stack from there at all and keep the OnboardingLayout responsible for stack maintenance.

igor-sirotin replied:
Agreed to keep this solution (at least for now) as the simplest, despite its inconsistency.

caybro commented on ui/main.qml:
```diff
@@ -520,6 +532,17 @@ StatusWindow {
                 }
             }
+
+            Component {
```
Same for the placement here...

caybro reviewed (commented):
In general looks good, just have some questions around the whole process/flow

igor-sirotin commented on ui/app/AppLayouts/Onboarding2/OnboardingFlow.qml:
```diff
@@ -533,4 +548,16 @@ OnboardingStackView {
+    Component {
+        id: reencryptingDatabaseScreen
+
+        SplashScreen {
```
I think I tried the `unwindToLoginScreen` and a direct login, but it didn't work. 
And before it required a restart as well, so I decided not try to do it all at once. We can tie it with the general `Logout` feature.

caybro replied:
Then I guess we should display that information to the user?

igor-sirotin replied:
Ok, pardon
Reencrpyiont is

caybro commented:
Oh, and one more comment/request: at least a simple QML test would be nice :wink:

micieslak commented on ui/main.qml:
```diff
@@ -520,6 +532,17 @@ StatusWindow {
+            Component {
+                id: convertingKeycardAccountPage
+
+                ConvertKeycardAccountPage {
```
What about the `tryAgain` handling there?

micieslak commented on ui/app/AppLayouts/Onboarding2/OnboardingFlow.qml:
```diff
@@ -533,4 +548,16 @@ OnboardingStackView {
     }
+
+    Component {
+        id: reencryptingDatabaseScreen
```
The back button should be suppressed on that page probably, by setting the marker property.

micieslak commented on ui/app/AppLayouts/Onboarding2/OnboardingFlow.qml:
```diff
@@ -133,6 +136,18 @@ OnboardingStackView {
+            if (root.reencryptingDatabase) {
+                root.push(reencryptingDatabaseScreen, StackView.Immediate)
+            } else {
+                root.pop(StackView.Immediate)
```
It's quite risky behavior, especially when we modify the stack internally and also from outside. We need to be sure that one we are going to pop is `reencryptingDatabaseScreen`. What if in the meantime some other page is pushed?

igor-sirotin replied:
This screen was removed now, according to the design 👌

micieslak reviewed (commented):
Looks ok in general, some code structure related questions.

igor-sirotin commented on src/app/modules/onboarding/post_onboarding/keycard_convert_account.nim:
```diff
@@ -0,0 +1,33 @@
+      currentPassword,
+      self.newPassword)
+
+  # WARNING: If biometrics were enabled, save the new passsword to it, instead of the pin
```
Yes, will implement now, was waiting for https://github.com/status-im/status-desktop/pull/17330 to be merged 👌

caybro commented:
The ui-tests are failing:
```
[2025-02-26T12:38:54.947Z] 3: QWARN  : QmlTests::UnknownTestFunc() file:///home/jenkins/workspace/s_linux_x86_64_tests-ui_PR-17285/storybook/qmlTests/tests/tst_OnboardingLayout.qml:163:5: Type OnboardingLayout unavailable 
[2025-02-26T12:38:54.947Z] 3:          property OnboardingLayout controlUnderTest: null 
[2025-02-26T12:38:54.947Z] 3:          ^
[2025-02-26T12:38:54.947Z] 3: QWARN  : QmlTests::UnknownTestFunc() file:///home/jenkins/workspace/s_linux_x86_64_tests-ui_PR-17285/ui/app/AppLayouts/Onboarding2/OnboardingLayout.qml:124:5: Type OnboardingFlow unavailable 
[2025-02-26T12:38:54.947Z] 3:          OnboardingFlow { 
[2025-02-26T12:38:54.947Z] 3:          ^
[2025-02-26T12:38:54.947Z] 3: QWARN  : QmlTests::UnknownTestFunc() file:///home/jenkins/workspace/s_linux_x86_64_tests-ui_PR-17285/ui/app/AppLayouts/Onboarding2/OnboardingFlow.qml:37:27: Duplicate property name 
[2025-02-26T12:38:54.947Z] 3:          required property var generateMnemonic 
[2025-02-26T12:38:54.947Z] 3:                                ^
[2025-02-26T12:38:54.947Z] 3/3 Test #3: QmlTests .........................***Failed   45.49 sec
```

igor-sirotin commented:
@caybro @micieslak @glitchminer 
I think I addressed all of the comments and updated the code as we agreed. Please let me know if I can improve anything else 🙂

caybro reviewed (commented):
Just some minor things

igor-sirotin commented on src/app/modules/onboarding/controller.nim:
```diff
@@ -116,6 +116,19 @@ proc init*(self: Controller) =
   self.connectionIds.add(handlerId)
 
+  handlerId = self.events.onWithUUID(SignalType.DBReEncryptionStarted.event) do(e: Args):
+    debug "reencryption process started"
```
@caybro Yes, I added it mostly to debug the freeze that happens for @glitchminer.
And in general, it doesn't hurt to have at all 🙂 It's a rare log, not private info

igor-sirotin commented on src/app/modules/onboarding/post_onboarding/save_biometrics_task.nim:
```diff
@@ -0,0 +1,23 @@
+    )
+
+proc run*(self: SaveBiometricsTask, accountsService: accounts_service.Service, onboardingModule: AccessInterface) =
+  debug "running post-onbaording SaveBiometricsTask"
```
The debug that I think is better to keep

igor-sirotin commented on ui/main.qml:
```diff
@@ -372,9 +374,13 @@ StatusWindow {
+                to: 1
+                duration: 30000
+                running: runningProgressAnimation
+                onFinished: {
```
@caybro Yes, but `onProgressChanged: { if (progress == 1) ...` didn't make any sense to me. 
This is exactly same as `finished` signal, isn't it?

caybro replied:
Not really... this (old) splash screen is launched in different ways; sometimes with the `progress` changing from `0..1`, sometimes not. CC @alexjba

caybro replied:
I'd just keep the old behavior, we don't actively use it anyways, do we?

alexjba replied:
The `finished` will be emitted only when the animation finishes naturally. So it really depends on how the splash screen is used externally. There's no short-circuit option when binding to the `finished` signal. This is why the progress was used to determine when to continue the app flow.

IMO using the `progress` property is more flexible and I'd keep the old behavior if the change is not a necessity.

On the other hand, looks to me like this change is safe since we don't short-circuit the splash screen.

igor-sirotin replied:
@caybro @alexjba thanks for insides! 
Initially, I made these change, because I actually relied on the `finished` signal here to do something.
But that something doesn't exist anymore, so I will rollback this change, at least for the sake of straight-forward file history 🙂

igor-sirotin commented on src/app/modules/onboarding/io_interface.nim:
```diff
@@ -94,6 +94,15 @@ method startKeycardFactoryReset*(self: AccessInterface) {.base.} =
+method onKeycardAccountConverted* (self: AccessInterface, success: bool) {.base.} =
+  raise newException(ValueError, "No implementation available")
+
+method requestSaveBiometrics*(self: AccessInterface, account: string, credential: string) {.base.} =
```
Seems that this works 👍

caybro reviewed (approved):
Good for me :)

micieslak reviewed (approved):
Nice work!

It looks good for now, some next steps we can consider after dropping the old onboarding, as we discussed and agreed.

Another thing we could consider later is the `OnboardingStore` API in order to make it potentially more granular.
- decomposing `finishOnboardingFlow` into dedicated methods with clearly specified parameters needed for given action
- I wonder if it's possible to decouple onboarding and keycard handling, not sure if it's possible from logical flow pov and how deeply it's coupled now

Another thing that we could review in order to introduce possibly handier pattern is async operations handling.

igor-sirotin commented on ui/app/AppLayouts/Onboarding2/pages/ConvertKeycardAccountPage.qml:
```diff
@@ -0,0 +1,214 @@
+OnboardingPage {
+    id: root
+
+    readonly property bool backAvailableHint: false
```
Hmm, but _for this page_ we never want to see the back button, that's why I added it here 🤔 
Should I still move it outside?

igor-sirotin commented:
@ulisesmac seems that `TestPairPendingContactRequest` is breaking for real in your PR 🙁

igor-sirotin commented:
@glitchminer I was able to reproduce the crash you mentioned. I thought this is because of `status-keycard-go` update, but doesn't seem to be the case. I'll check what's wrong there.
